### PR TITLE
Refactor flow tests

### LINKED ISSUE #10: Refactor flow tests to use page objects
#### As a developer I want my flow tests to be written using the object-oriented [page object pattern](http://martinfowler.com/bliki/PageObject.html).

__Story__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the ae_page_objects readme which will be an extremely
helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Acceptance criteria__:
- [x] The flow tests have been refactored to use page objects.

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://github.com/appfolio/ae_page_objects
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

__Dependencies__:
- Delete Images
- Images by Tag

### PR DESCRIPTION
**Pull request by almgong**
Closes #10

### CODE REVIEW

File: test/page_objects/images/index_page.rb [hunk 1]
```diff
@@ -3,19 +3,21 @@ module Images
     class IndexPage < PageObjects::Document
       path :images
 
-      collection :images, locator: '#TODO', item_locator: '#TODO', contains: ImageCard do
+      collection :images, locator: '.js-image-card-container', item_locator: '.card', contains: ImageCard do
```
Comment (kkajla12) [inline test/page_objects/images/index_page.rb]: I think it's fine to leave `.card` as the `item_locator` for our purposes here, but just something to think about: You're indirectly tying this page object to a particular class that has styling associated with it. The upside of using a 'js' class whose sole purpose is to be for selecting a specific element in javascript or page object code is that it is understood that the class is meant specifically for those purposes and doesn't have any reason to be changed unless the corresponding code is changed as well. In this case, if you were decide to use a different css class to style your cards, you would need to change the page object as well.

File: test/page_objects/images/show_page.rb [hunk 2]
```diff
@@ -4,11 +4,11 @@ class ShowPage < PageObjects::Document
       path :image
 
       def image_url
-        #TODO
+        node.find('img')[:src]
```
Comment (kkajla12) [inline test/page_objects/images/show_page.rb]: Maybe this is something you could clean up by adding it as an `element` on this page object?

`element :image, locator: `img`

Then you can say something like `image.node[:src]` here to return the url. It's a bit more idiomatic and you're letting page objects do most of the work for you.

File: test/page_objects/images/show_page.rb [hunk 3]
```diff
@@ -4,11 +4,11 @@ class ShowPage < PageObjects::Document
...
       end
 
       def tags
-        #TODO
+        node.find('.js-tags').text.split
```
Comment (kkajla12) [inline test/page_objects/images/show_page.rb]: Refer to https://github.com/AppFolioOnboarding/image-sharer-almgong/pull/30/commits/324a4e0a542e7930a2720c2581bee96452ad849b#r109209190

Here you could do something like:

`element :tag_list, locator: '.js-tags', item_locator: '.js-class'`

and then this method can do something like:

`tag_list.map(&:text)`

File: test/flows/images_crud_test.rb [hunk 4]
```diff
@@ -45,14 +45,12 @@ class ImagesCrudTest < FlowTestCase
       image.url == ugly_cat_url
     end
     image_show_page = image_to_delete.view!
-
     image_show_page.delete do |confirm_dialog|
-      assert_equal 'Are you sure?', confirm_dialog.text
+      assert_equal 'Are you sure you want to delete?', confirm_dialog.text
```
Comment (kkajla12) [inline test/flows/images_crud_test.rb]: Love the more specific error message! :)

File: Gemfile [hunk 5]
```diff
@@ -1,6 +1,9 @@
 source 'https://rubygems.org'
 ruby '2.3.2'
 
+# explicitly use version of rubocop
+gem 'rubocop', '~> 0.40.0', require: false
```
Comment (kkajla12) [inline Gemfile]: What was the reason for this?

Comment (almgong) [inline Gemfile]: I am not 100% certain on what had happened, but after checking out the provided branch and running `bundle install`, the version of rubocop was updated to something around `0.48.0`. It led to a ton of failing rubocop errors (a handful of cryptic ones as well). I decided at this time that the effort and time spent on going back and fixing all of these new rubocop issues was not worth the time, so I specified the original version of rubocop that I had been using since the beginning of ropes.

Do you think that I should undo this and use the latest version of rubocop?

File: Gemfile.lock [hunk 6]
```diff
@@ -2,93 +2,95 @@ GEM
   remote: https://rubygems.org/
   remote: https://rails-assets.org/
   specs:
-    actioncable (5.0.0.1)
-      actionpack (= 5.0.0.1)
-      nio4r (~> 1.2)
+    actioncable (5.0.2)
```
Comment (almgong) [inline Gemfile.lock]: This was also a result of running `bundle install` after checking out from the provided branch. I think it's necessary for this PR because without it, the project may not run the flow tests as expected.

However, I think it's best (as you said later on) to move this to a separate commit.

Comment (almgong) [inline Gemfile.lock]: ***EDIT

Actually, I think that perhaps somewhere down the line `bundle update` must have been invoked. I looked through the original branch I checked out of and it seems like there isn't a reason that these got updated. 

I will checkout from the branch again and re-apply the other to confirm.

File: Gemfile [hunk 7]
```diff
@@ -61,7 +61,7 @@ end
 group :test do
   gem 'ae_page_objects'
   gem 'capybara'
-  gem 'selenium-webdriver'
+  gem 'selenium-webdriver', '~> 2.53'
```
Comment (almgong) [inline Gemfile]: I understand what you are getting at. I think things like this would not normally be part of a PR that deals with adding functionality (since you would more likely be adding gems than downgrading existing ones anyways). I guess for the sake of this issue on ropes and enabling someone to *more easily run the flow tests after forking this repo, I will keep the downgrading in this PR. 

I will, however, split this and have the downgrading of selenium in its own commit.

File: Gemfile.lock (hunk 6 above)
Comment (almgong) [inline Gemfile.lock]: You were right! These changes to Gemfile.lock have been removed.

File: test/page_objects/images/index_page.rb (hunk 1 above)
Comment (almgong) [inline test/page_objects/images/index_page.rb]: That is very true. I will update the collection item_locator to have a `js` class like `.js-card`.

File: Gemfile (hunk 5 above)
Comment (kkajla12) [inline Gemfile]: I feel like you should be using the latest version of rubocop. Maybe that (and any changes to fix rubocop errors) can be resolved in a separate PR though. This is fine for now.

File: test/page_objects/images/image_card.rb [hunk 8]
```diff
@@ -6,7 +6,7 @@ def url
       end
 
       def tags
-        #TODO
+        node.find('.card-text').text.split
```
Comment (kkajla12) [inline test/page_objects/images/image_card.rb]: You could do the same thing with `tag_list` here (make it a `collection`) so that you can have a clean call to `map`

**Comment by kkajla12** [comment]
Left one or two more comments. Once you've looked into those, I think you're good to merge! Nice job!

File: Gemfile (hunk 5 above)
Comment (almgong) [inline Gemfile]: This line actually got removed when I "undid" the result of invoking `bundle update`. And because of that I also did not change rubocop versions. However I will update all of my packages outside of this PR and work on that at a later time.

File: test/page_objects/images/image_card.rb (hunk 8 above)
Comment (almgong) [inline test/page_objects/images/image_card.rb]: I'll do the same with the `url` function above, and define something like `element :image, locator: 'img'` to use in `image.node[:src]`
